src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,330 @@
+    private array $unset = [];
+
+    private bool $entirePackedArrayRead = false;
+
+    private CodecLibrary $codecLibrary;
+
+    /**
+     * Deep clone this lazy array.
+     */
+    public function __clone()
+    {
+        $this->bson = clone $this->bson;
+
+        foreach ($this->set as $key => $value) {
+            $this->set[$key] = recursive_copy($value);
+        }
+    }
+
+    /** @param PackedArray<TValue>|list<TValue>|null $input */
+    public function __construct($input = null, ?CodecLibrary $codecLibrary = null)
+    {
+        if ($input === null) {
+            $this->bson = PackedArray::fromPHP([]);
+        } elseif ($input instanceof PackedArray) {
+            $this->bson = $input;
+        } elseif (is_array($input)) {
+            if (! array_is_list($input)) {
+                // Check if all keys are numeric
+                foreach ($input as $key => $value) {
+                    if (! is_numeric($key)) {
+                        throw InvalidArgumentException::invalidType('input', $input, [PackedArray::class, 'array', 'null']);
+                    }
+                }
+            }
+
+            $this->bson = PackedArray::fromPHP([]);
+            $this->set = $input;
+            $this->exists = array_map(
+                /** @param TValue $value */
+                fn ($value): bool => true,
+                $this->set,
+            );
+        } else {
+            throw InvalidArgumentException::invalidType('input', $input, [PackedArray::class, 'array', 'null']);
+        }
+
+        $this->codecLibrary = $codecLibrary ?? new LazyBSONCodecLibrary();
+    }
+
+    /** @return array{bson: PackedArray<TValue>, set: array<int, TValue>, unset: array<int, true>, codecLibrary: CodecLibrary} */
+    public function __serialize(): array
+    {
+        return [
+            'bson' => $this->bson,
+            'set' => $this->set,
+            'unset' => $this->unset,
+            'codecLibrary' => $this->codecLibrary,
+        ];
+    }
+
+    /** @param array{bson: PackedArray<TValue>, set: array<int, TValue>, unset: array<int, true>, codecLibrary: CodecLibrary} $data */
+    public function __unserialize(array $data): void
+    {
+        $this->bson = $data['bson'];
+        $this->set = $data['set'];
+        $this->unset = $data['unset'];
+        $this->codecLibrary = $data['codecLibrary'];
+
+        $this->exists = array_map(
+        /** @param TValue $value */
+            fn ($value): bool => true,
+            $this->set,
+        );
+
+        foreach ($this->unset as $index => $unused) {
+            $this->exists[$index] = false;
+        }
+    }
+
+    public function count(): int
+    {
+        $this->readEntirePackedArray();
+
+        return count(array_filter($this->exists));
+    }
+
+    /** @return ListIterator<TValue> */
+    public function getIterator(): ListIterator
+    {
+        // Sort keys to ensure they are in ascending order
+        ksort($this->set, SORT_NUMERIC);
```

Comment:
See my comment above. I think sorting here might create differing behavior between BSONArray and LazyBSONArray. Deferring any reordering until serialization time seems more consistent.

If these Lazy classes aren't going to implement BSON\Serializable and _must_ rely on the codec for encoding, then I don't think we need to sort here. `LazyBSONArrayCodec::encode()` already disregards keys by using `$return[]` to build up the list used for constructing a PackedArray. This sorting appears to be here for the benefit of `jsonSerialize()`, which currently only uses `iterator_to_array()`.

If the user constructs a LazyBSONArray with an out-of-sequence PHP array (e.g. `[1 => 'foo', 0 => 'bar']`, then I'd expect iteration to return the values "foo" and "bar" in sequence based on my previous experience with BSONArray, ArrayObject, and basic arrays. Moreover, BSONArray's BSON and JSON encoding would yield the sequence `["foo", "bar"]`. That's the existing behavior users get with BSONArray.

I think the common rule with both classes should be that "keys do not dictate list order" if we aim to be consistent.

And maybe that's the root question here: are we prioritizing consistency with BSONArray and PHP arrays, or are we trying to do something new with LazyBSONArray?

The current implementation of this class feels more like a mutable version of PackedArray, and I think that's the reason we're struggling so much. [SplDoublyLinkedList](https://www.php.net/manual/en/class.spldoublylinkedlist.php) can function as a mutable PackedArray, but trying to provide the sloppy ArrayObject or array API is inviting all sorts of edge cases.